### PR TITLE
Write metadata cache data to mappings _meta with refresh time update

### PR DESCRIPTION
**Pull request by seankao-az**
#### Metadata Cache Writer
For the most part, same as
* #744 

In addition to the regular metadata storage using `FlintIndexMetadataService`, we're dual-writing additional fields, defined by `FlintMetadataCache`, to the index mappings `_meta` field. It's intended for frontend users to access some crucial metadata for an index quickly without invoking another backend API call.

This PR adds such fields for all indexes, if the spark config `spark.flint.metadataCacheWrite.enabled` is set to true.
* `_meta.properties.metadataCacheVersion`: "1.0"
* `_meta.properties.refreshInterval`: Integer. Refresh interval of an index measured in seconds. This field is added only if index refresh type is auto refresh and refresh_interval is set
* `_meta.properties.sourceTables`: Array of Strings. For now, it's mocked data. Update coming in later PR.
* `_meta.properties.lastRefreshTime`: Long. Timestamp in milliseconds when last refresh happened. This field is added only if index already gets refreshed at least once

#### Last Refresh Time
Added two new fields in `FlintMetadataLogEntry` and bumped version of its json doc from 1.0 to 1.1 (because adding new field but not changing existing fields)
* lastRefreshStartTime: Long. Timestamp when last refresh started
* lastRefreshCompleteTime: Long. Timestamp when last refresh completed

These are accurate *only* for manual refresh (full, incremental) and external scheduler for auto refresh.
For internal scheduler, the `jobStartTime` (or `createTime` in `FlintMetadataLogEntry`) is used to track streaming job start time.

I'm not reusing `createTime` because they should be updated at different times.
For createTime (for internal scheduler) it's during `refreshIndex`, `recoverIndex`, `updateIndexManualToAuto`
But for lastRefreshStartTime and lastRefreshCompleteTime (for manual refresh and external scheduler) it's only updated in `refreshIndex`

#### End-to-End Test
Tests performed in my test cluster
<details>
<summary>index (full refresh mode) created without `spark.flint.metadataCacheWrite.enabled`</summary>

Checking the `_meta.properties` field: No fields for metadata cache is added
```
GET flint_myglue_test_default_http_logs_cv_full_no_cache_index/_mappings

{
  "flint_myglue_test_default_http_logs_cv_full_no_cache_index": {
    "mappings": {
      "_meta": {
        "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfZnVsbF9ub19jYWNoZV9pbmRleA==",
        "kind": "covering",
        "indexedColumns": [
          {
            "columnType": "timestamp",
            "columnName": "@timestamp"
          },
          {
            "columnType": "string",
            "columnName": "request"
          },
          {
            "columnType": "int",
            "columnName": "size"
          },
          {
            "columnType": "string",
            "columnName": "clientip"
          },
          {
            "columnType": "int",
            "columnName": "status"
          }
        ],
        "name": "cv_full_no_cache",
        "options": {
          "auto_refresh": "false",
          "incremental_refresh": "false"
        },
        "source": "myglue_test.default.http_logs",
        "version": "0.6.0",
        "properties": {
          "filterCondition": "status = 403",
          "env": {
            "SERVERLESS_EMR_VIRTUAL_CLUSTER_ID": "****",
            "SERVERLESS_EMR_JOB_ID": "****"
          }
        }
      },
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "strict_date_optional_time_nanos"
        },
        "clientip": {
          "type": "keyword"
        },
        "request": {
          "type": "keyword"
        },
        "size": {
          "type": "integer"
        },
        "status": {
          "type": "integer"
        }
      }
    }
  }
}
```
</details>

<details>
<summary>index (full refresh mode) created with `spark.flint.metadataCacheWrite.enabled` set to true </summary>

Checking the `_meta.properties` field:
* metadataCacheVersion is set to 1.0
* sourceTables is set to mocked array data
```
GET flint_myglue_test_default_http_logs_cv_full_with_cache_index/_mappings

{
  "flint_myglue_test_default_http_logs_cv_full_with_cache_index": {
    "mappings": {
      "_meta": {
        "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfZnVsbF93aXRoX2NhY2hlX2luZGV4",
        "kind": "covering",
        "indexedColumns": [
          {
            "columnType": "timestamp",
            "columnName": "@timestamp"
          },
          {
            "columnType": "string",
            "columnName": "request"
          },
          {
            "columnType": "int",
            "columnName": "size"
          },
          {
            "columnType": "string",
            "columnName": "clientip"
          },
          {
            "columnType": "int",
            "columnName": "status"
          }
        ],
        "name": "cv_full_with_cache",
        "options": {
          "auto_refresh": "false",
          "incremental_refresh": "false"
        },
        "source": "myglue_test.default.http_logs",
        "version": "0.6.0",
        "properties": {
          "sourceTables": [
            "dataSourceName.default.logGroups(logGroupIdentifier:['arn:aws:logs:us-east-1:123456:test-llt-xa', 'arn:aws:logs:us-east-1:123456:sample-lg-1'])"
          ],
          "filterCondition": "status = 403",
          "metadataCacheVersion": "1.0",
          "env": {
            "SERVERLESS_EMR_VIRTUAL_CLUSTER_ID": "****",
            "SERVERLESS_EMR_JOB_ID": "****"
          }
        }
      },
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "strict_date_optional_time_nanos"
        },
        "clientip": {
          "type": "keyword"
        },
        "request": {
          "type": "keyword"
        },
        "size": {
          "type": "integer"
        },
        "status": {
          "type": "integer"
        }
      }
    }
  }
}
```
</details>

<details>
<summary>Triggering a full refresh</summary>

Check that `lastRefreshStartTime` is updated when index (full refresh mode) enters refreshing
```
GET .query_execution_request_myglue_test/_search

      {
        "_index": ".query_execution_request_myglue_test",
        "_id": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfZnVsbF93aXRoX2NhY2hlX2luZGV4",
        "_score": 1.0,
        "_source": {
          "version": "1.1",
          "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfZnVsbF93aXRoX2NhY2hlX2luZGV4",
          "type": "flintindexstate",
          "state": "refreshing",
          "applicationId": "****",
          "jobId": "****",
          "dataSourceName": "myglue_test",
          "jobStartTime": 1730241864568,
          "lastRefreshStartTime": 1730241864568,
          "lastRefreshCompleteTime": 0,
          "lastUpdateTime": 1730241864691,
          "error": ""
        }
      }
```

And `lastRefreshCompleteTime` updated when refresh is done
```
GET .query_execution_request_myglue_test/_search

      {
        "_index": ".query_execution_request_myglue_test",
        "_id": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfZnVsbF93aXRoX2NhY2hlX2luZGV4",
        "_score": 1.0,
        "_source": {
          "version": "1.1",
          "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfZnVsbF93aXRoX2NhY2hlX2luZGV4",
          "type": "flintindexstate",
          "state": "active",
          "applicationId": "****",
          "jobId": "****",
          "dataSourceName": "myglue_test",
          "jobStartTime": 1730241864568,
          "lastRefreshStartTime": 1730241864568,
          "lastRefreshCompleteTime": 1730241882050,
          "lastUpdateTime": 1730241882107,
          "error": ""
        }
      }
```

The `_meta.properties.lastRefreshTime` is added as well when `spark.flint.metadataCacheWrite.enabled` set to true
```

{
  "flint_myglue_test_default_http_logs_cv_full_with_cache_index": {
    "mappings": {
      "_meta": {
        "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfZnVsbF93aXRoX2NhY2hlX2luZGV4",
        "kind": "covering",
        "indexedColumns": [
          {
            "columnType": "timestamp",
            "columnName": "@timestamp"
          },
          {
            "columnType": "string",
            "columnName": "request"
          },
          {
            "columnType": "int",
            "columnName": "size"
          },
          {
            "columnType": "string",
            "columnName": "clientip"
          },
          {
            "columnType": "int",
            "columnName": "status"
          }
        ],
        "name": "cv_full_with_cache",
        "options": {
          "auto_refresh": "false",
          "incremental_refresh": "false"
        },
        "source": "myglue_test.default.http_logs",
        "version": "0.6.0",
        "properties": {
          "sourceTables": [
            "dataSourceName.default.logGroups(logGroupIdentifier:['arn:aws:logs:us-east-1:123456:test-llt-xa', 'arn:aws:logs:us-east-1:123456:sample-lg-1'])"
          ],
          "filterCondition": "status = 403",
          "metadataCacheVersion": "1.0",
          "lastRefreshTime": 1730241882050,
          "env": {
            "SERVERLESS_EMR_VIRTUAL_CLUSTER_ID": "****",
            "SERVERLESS_EMR_JOB_ID": "****"
          }
        }
      },
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "strict_date_optional_time_nanos"
        },
        "clientip": {
          "type": "keyword"
        },
        "request": {
          "type": "keyword"
        },
        "size": {
          "type": "integer"
        },
        "status": {
          "type": "integer"
        }
      }
    }
  }
}
```

</details>

<details>
<summary>Test with auto refresh with external scheduler with cache write enabled</summary>

`_meta.properties.refreshInterval` is filled with number (unit: seconds) as expected
```
{
  "flint_myglue_test_default_http_logs_cv_auto_with_cache_index": {
    "mappings": {
      "_meta": {
        "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfYXV0b193aXRoX2NhY2hlX2luZGV4",
        "kind": "covering",
        "indexedColumns": [
          {
            "columnType": "timestamp",
            "columnName": "@timestamp"
          },
          {
            "columnType": "string",
            "columnName": "request"
          },
          {
            "columnType": "int",
            "columnName": "size"
          },
          {
            "columnType": "string",
            "columnName": "clientip"
          },
          {
            "columnType": "int",
            "columnName": "status"
          }
        ],
        "name": "cv_auto_with_cache",
        "options": {
          "auto_refresh": "true",
          "refresh_interval": "5 Minutes",
          "scheduler_mode": "external",
          "incremental_refresh": "false",
          "checkpoint_location": "s3://flint-dev-seankao/checkpoints/metadata_cache_test_auto_2/"
        },
        "source": "myglue_test.default.http_logs",
        "version": "0.6.0",
        "properties": {
          "refreshInterval": 300,
          "sourceTables": [
            "dataSourceName.default.logGroups(logGroupIdentifier:['arn:aws:logs:us-east-1:123456:test-llt-xa', 'arn:aws:logs:us-east-1:123456:sample-lg-1'])"
          ],
          "filterCondition": "status = 403",
          "metadataCacheVersion": "1.0",
          "env": {
            "SERVERLESS_EMR_VIRTUAL_CLUSTER_ID": "****",
            "SERVERLESS_EMR_JOB_ID": "****"
          }
        }
      },
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "strict_date_optional_time_nanos"
        },
        "clientip": {
          "type": "keyword"
        },
        "request": {
          "type": "keyword"
        },
        "size": {
          "type": "integer"
        },
        "status": {
          "type": "integer"
        }
      }
    }
  }
}
```

lastRefreshStartTime and lastRefreshCompleteTime are updated accordingly
```
      {
        "_index": ".query_execution_request_myglue_test",
        "_id": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfYXV0b193aXRoX2NhY2hlX2luZGV4",
        "_score": 0.7801585,
        "_source": {
          "version": "1.1",
          "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfYXV0b193aXRoX2NhY2hlX2luZGV4",
          "type": "flintindexstate",
          "state": "active",
          "applicationId": "****",
          "jobId": "****",
          "dataSourceName": "myglue_test",
          "jobStartTime": 1730244879973,
          "lastRefreshStartTime": 1730244879973,
          "lastRefreshCompleteTime": 1730244901683,
          "lastUpdateTime": 1730244901696,
          "error": ""
        }
      }
```

lastRefreshTime also added to `_meta.properties`
```
{
  "flint_myglue_test_default_http_logs_cv_auto_with_cache_index": {
    "mappings": {
      "_meta": {
        "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfYXV0b193aXRoX2NhY2hlX2luZGV4",
        "kind": "covering",
        "indexedColumns": [
          {
            "columnType": "timestamp",
            "columnName": "@timestamp"
          },
          {
            "columnType": "string",
            "columnName": "request"
          },
          {
            "columnType": "int",
            "columnName": "size"
          },
          {
            "columnType": "string",
            "columnName": "clientip"
          },
          {
            "columnType": "int",
            "columnName": "status"
          }
        ],
        "name": "cv_auto_with_cache",
        "options": {
          "auto_refresh": "true",
          "refresh_interval": "5 Minutes",
          "scheduler_mode": "external",
          "incremental_refresh": "false",
          "checkpoint_location": "s3://flint-dev-seankao/checkpoints/metadata_cache_test_auto_2/"
        },
        "source": "myglue_test.default.http_logs",
        "version": "0.6.0",
        "properties": {
          "refreshInterval": 300,
          "sourceTables": [
            "dataSourceName.default.logGroups(logGroupIdentifier:['arn:aws:logs:us-east-1:123456:test-llt-xa', 'arn:aws:logs:us-east-1:123456:sample-lg-1'])"
          ],
          "filterCondition": "status = 403",
          "metadataCacheVersion": "1.0",
          "lastRefreshTime": 1730245374777,
          "env": {
            "SERVERLESS_EMR_VIRTUAL_CLUSTER_ID": "****",
            "SERVERLESS_EMR_JOB_ID": "****"
          }
        }
      },
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "strict_date_optional_time_nanos"
        },
        "clientip": {
          "type": "keyword"
        },
        "request": {
          "type": "keyword"
        },
        "size": {
          "type": "integer"
        },
        "status": {
          "type": "integer"
        }
      }
    }
  }
}
```
</details>

<details>
<summary>Test for auto refresh with internal scheduler</summary>

jobStartTime is updated for streaming job, but lastRefreshStart/CompleteTime isn't updated, as expected
```
GET /.query_execution_request_myglue_test/_search

      {
        "_index": ".query_execution_request_myglue_test",
        "_id": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfYXV0b19pbnRlcm5hbF93aXRoX2NhY2hlX2luZGV4",
        "_score": 0.90223897,
        "_source": {
          "version": "1.1",
          "latestId": "ZmxpbnRfbXlnbHVlX3Rlc3RfZGVmYXVsdF9odHRwX2xvZ3NfY3ZfYXV0b19pbnRlcm5hbF93aXRoX2NhY2hlX2luZGV4",
          "type": "flintindexstate",
          "state": "refreshing",
          "applicationId": "****",
          "jobId": "****",
          "dataSourceName": "myglue_test",
          "jobStartTime": 1730244344917,
          "lastRefreshStartTime": 0,
          "lastRefreshCompleteTime": 0,
          "lastUpdateTime": 1730244726126,
          "error": ""
        }
      }
```
</details>

### Related Issues
* #746 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/scheduler/util/IntervalSchedulerParser.java [hunk 1]
```diff
@@ -18,21 +18,30 @@
...
-     * Parses a schedule string into an IntervalSchedule.
+     * Parses a schedule string into an integer in milliseconds.
      *
      * @param scheduleStr the schedule string to parse
-     * @return the parsed IntervalSchedule
+     * @return the parsed integer
      * @throws IllegalArgumentException if the schedule string is invalid
      */
-    public static IntervalSchedule parse(String scheduleStr) {
+    public static Long parseMillis(String scheduleStr) {
```
Comment (ykmr1224) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/scheduler/util/IntervalSchedulerParser.java:27]: ParseMillis sounds like parsing millisec string. Should we call it `parseAndConvertToMillis`?

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala [hunk 2]
```diff
@@ -520,6 +513,62 @@ class FlintSpark(val spark: SparkSession) extends FlintSparkTransactionSupport w
     updatedOptions.isExternalSchedulerEnabled() != originalOptions.isExternalSchedulerEnabled()
   }
 
+  /**
+   * Handles refresh for refresh mode AUTO, which is used exclusively by auto refresh index with
+   * internal scheduler.
+   */
+  private def refreshIndexAuto(
```
[the remark this answers is not in the document]
Comment (seankao-az) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala:520]: for now only track `lastRefreshStartTime` and `lastRefreshCompleteTime` for manual refresh and auto refresh with external scheduler.

for streaming job, we use `createTime` to track the streaming job start time.
there's no tracking timestamp for each micro batch update yet, so updating the 2 timestamp could be misleading.

[the remark this answers is not in the document]
Comment (seankao-az) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala:520]: I'll add some comment

**Comment by seankao-az** [comment]
Note to any reviewer if curious, the force push only amended commit [2f58f56](https://github.com/opensearch-project/opensearch-spark/pull/805/commits/2f58f567959ce68257ffa033a4878d61d2e24e2e) and nothing else

**Review by ykmr1224 — COMMENTED** [review]
Why do we call it metadata `Cache`? I was not quite sure the indication of `cache`.

**Comment by seankao-az** [comment]
i do welcome a better name... was kind of struggling to come up with a name.
So the main use case for this is when using custom index metadata and metadata log storage, these metadata aren't available in OpenSearch index. And some frontend use case need access to these data without making query to backend executed in spark. I take it as a read cache for such metadata.

**Comment by seankao-az** [comment]
Will deploy and perform end-to-end test tonight before merging

**Comment by seankao-az** [comment]
Updated description with test results

**Comment by seankao-az** [comment]
Backport blocked by: https://github.com/opensearch-project/opensearch-spark/pull/836
retrigger backport once revert pr is merged

**Pull request merged**